python/tank/authentication/shotgun_shared/saml2_sso.py

```diff
@@ -582,17 +575,150 @@ def on_sso_disable_renewal(self, event):
 ################################################################################
 
 
-def _decode_cookies(app_cookies):
-    sg_cookies = []
-    if app_cookies:
-        sg_cookies = json.loads(base64.b64decode(app_cookies))
-        sg_cookies = [cookie.encode("utf-8") for cookie in sg_cookies]
-    return sg_cookies
+def _decode_cookies(encoded_cookies):
+    """
+    Extract the cookies from a base64 encoded string.
+
+    :param encoded_cookies: An encoded string representing the cookie jar.
+
+    :returns: A SimpleCookie containing all the cookies.
+    """
+    cookies = SimpleCookie()
+    if encoded_cookies:
+        try:
+            decoded_cookies = base64.b64decode(encoded_cookies)
+            cookies.load(decoded_cookies)
+        except TypeError, e:
+            log.error("Unable to decode the cookies: %s" % e.message)
+    return cookies
+
+
+def _encode_cookies(cookies):
+    """
+    Extract the cookies from a base64 encoded string.
+
+    :param cookies: A Cookie.SimpleCookie instance representing the cookie jar.
+
+    :returns: An encoded string representing the cookie jar.
+    """
+    encoded_cookies = base64.b64encode(cookies.output())
+    return encoded_cookies
+
+
+def _get_shotgun_user_id(cookies):
+    """
+    Returns the id of the user in the shotgun instance, based on the cookies.
+
+    :param cookies: A Cookie.SimpleCookie instance representing the cookie jar.
+
+    :returns: A string user id value, or None.
+    """
+    user_id = None
+    for cookie in cookies:
+        if cookie.startswith("shotgun_sso_session_userid_u"):
+            if user_id is not None:
+                # Should we find multiple cookies with the same prefix, it means
+                # that we are using cookies from a multi-session environment. We
+                # have no way to identify the proper user id in the lot.
+                raise Exception('Multi-session not supported')
```

Comment:
To be honest, the current code was sloppy on my part... I wrote it in quickly, just to have it there. But forgot to turn it into a first class exception citizen.